_posts/2025-06-24-kafka-queue.md

```diff
@@ -0,0 +1,703 @@
+
+On the other hand, a Kafka topic can be consumed by multiple consumer groups as it is an immutable log and reading of records does not result in removal. However, workload distribution cannot be achieved alone with consumers. A topic has to be partitioned in order to spread the workload across a consumer group. It also provides an ordering mechanism by using record keys.
+
+The Queues for Kafka feature allows you to use a Kafka topic like a traditional message queue and increase parallelism of message processing beyond the number of partitions. It provides queue-like semantics through a new consumer group called <b>share group</b> to enable use cases for traditional message queue. This new type of group gives more fine-grained control on message acknowledgement and retries.
+
+The key difference between share group and regular consumer group is how partitions get assigned to consumer members. With regular consumer groups, each partition is exclusively assigned to a single member of the consumer group. Therefore users typically can have as many consumer members as the number of partitions to maximise the parallelism in message processing. Moreover, due to this, users tend to over partition their topics in order to cope with peak loads that may only happen sometimes. However, share groups balance partitions between all members of a share group, allowing multiple consumer members to fetch from the same partition. So users can have more consumers than the number of partitions, further increasing the parallelism and they do not need to over partition their topics, but can just scale up and down their consumers to cope with the peak loads. When share group members consume from the same partition, each record on that partition is still only read by one consumer in the group.
+
+The share group does not support exactly-once semantics and does not provide ordering guarantee as multiple consumers can fetch records from the same partition. However, these features may potentially be added in the future. The dead letter queue is currently not supported either but there is already an open [KIP](https://cwiki.apache.org/confluence/display/KAFKA/KIP-1191%3A+Dead-letter+queues+for+share+groups) for it.
+
+### Comparing share and consumer groups with Strimzi
+
+I have set up a single node cluster on my machine using Strimzi’s [Quickstart](https://strimzi.io/quickstarts/) and created a topic called `kafka-queue-topic` with 2 partitions:
+```sh
+$ cat << EOF | kubectl create -n kafka -f -
+apiVersion: kafka.strimzi.io/v1beta1
+kind: KafkaTopic
+metadata:
+  name: kafka-queue-topic
+  labels:
+    strimzi.io/cluster: "my-cluster"
+spec:
+  partitions: 2
+  replicas: 1
+EOF
+```
+
+Since this feature is still in early access, as of Apache Kafka 4.0.0, you need to explicitly enable it by setting the following:
```

Comment:
Technically, there is nothing wrong with the order as you have it. But somehow I personally would first talk about preparing the Kafka cluster and only then create the topic ... just saying, I know it might be individual.